core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java

```diff
@@ -107,106 +96,129 @@
+   * @param subject the parent subject, set to null if not present
+   * @param alluxioConf Alluxio's configuration
+   * @return the context
    */
-  public static FileSystemContext get(Subject subject) {
-    synchronized (CONTEXT_CACHE_LOCK) {
-      FileSystemContext ctx = CONTEXT_CACHE.computeIfAbsent(subject, FileSystemContext::create);
-      ctx.mRefCount++;
-      return ctx;
+  public static FileSystemContext create(Subject subject,
+      @Nullable AlluxioConfiguration alluxioConf) {
+    AlluxioProperties props = null;
+    if (alluxioConf != null) {
+      props = alluxioConf.getProperties();
     }
+    return create(subject, props);
   }
 
   /**
    * @param subject the parent subject, set to null if not present
+   * @param props Alluxio's configuration properties
    * @return the context
    */
-  private static FileSystemContext create(Subject subject) {
-    FileSystemContext context = new FileSystemContext(subject);
-    context.init(MasterInquireClient.Factory.create(), Configuration.global());
+  public static FileSystemContext create(Subject subject, @Nullable AlluxioProperties props) {
+    FileSystemContext context = new FileSystemContext(subject, props);
+    context.init(MasterInquireClient.Factory.create(context.getClientContext().getConf()));
     return context;
   }
 
   /**
-   * This method is provided for testing, use the {@link FileSystemContext#get} methods. The
+   * @param clientContext the {@link alluxio.ClientContext} containing the subject and configuration
+   * @return the {@link alluxio.client.file.FileSystemContext}
+   */
+  public static FileSystemContext create(ClientContext clientContext) {
+    Preconditions.checkNotNull(clientContext);
+    FileSystemContext ctx = new FileSystemContext(clientContext);
+    ctx.init(MasterInquireClient.Factory.create(clientContext.getConf()));
+    return ctx;
+  }
+
+  /**
+   * This method is provided for testing, use the {@link FileSystemContext#create} methods. The
    * returned context object will not be cached automatically.
    *
    * @param subject the parent subject, set to null if not present
    * @param masterInquireClient the client to use for determining the master; note that if the
    *        context is reset, this client will be replaced with a new masterInquireClient based on
-   *        global configuration
+   *        the original configuration.
+   * @param alluxioConf Alluxio's configuration
    * @return the context
    */
   @VisibleForTesting
-  public static FileSystemContext create(Subject subject, MasterInquireClient masterInquireClient) {
-    FileSystemContext context = new FileSystemContext(subject);
-    context.init(masterInquireClient, Configuration.global());
-    synchronized (CONTEXT_CACHE_LOCK) { // Not necessary, for code consistency
-      context.mRefCount++;
+  public static FileSystemContext create(Subject subject, MasterInquireClient masterInquireClient,
```

Comment:
I think this constructor is only used in testing code to inject a specific master inquire client. Otherwise it's not used.

I just checked and it's only used inside the `MultiProcessCluster`

Reply:
So I am revisiting this in design and I think it might make sense to include a `MasterInquireClient` in the client context. It also falls in line with the design of the class for keeping all information which is needed to connect within the `ClientContext` as well

Reply:
Making quite a few changes. See how `ClientContext` is used within `AbstractClient` and it's use within `MasterClientConfig`

Reply:
https://github.com/Alluxio/alluxio/pull/8277/commits/c36849ecb337229b05ae0b9aee2bf60762b30c0b